app/models/miq_provision_virt_workflow.rb

```diff
@@ -216,15 +216,19 @@ def available_vlans_and_hosts(options = {})
       rails_logger('allowed_vlans', 0)
       # TODO: Use Active Record to preload this data?
       MiqPreloader.preload(hosts, :switches => :lans)
-      hosts.each do |h|
-        h.lans.each { |l| vlans[l.name] = l.name unless l.switch.shared? }
-      end
+      load_allowed_vlans(hosts, vlans)
       rails_logger('allowed_vlans', 1)
     end
 
     return vlans, hosts
   end
 
+  def load_allowed_vlans(hosts, vlans)
+    hosts.each do |h|
+      h.lans.each { |l| vlans[l.name] = l.name if @vlan_options[:dvs] || !l.switch.shared? }
```

Comment:
Looking at this more I don't think you need to touch this at all, its just moving code around and its just three lines so I don't think it helps readability, unless you're planning on overriding the `load_allowed_vlans` method in a subclass I think you should just leave it the way it is.